js.go

```diff
@@ -838,27 +837,34 @@ type jsSub struct {
 	active bool
 	fc     bool
 	cmeta  string
-	fcs    map[uint64]string
+	fcr    string
+	fcd    uint64
 }
 
-func (jsi *jsSub) unsubscribe(drainMode bool) error {
-	jsi.mu.Lock()
-	durable, attached := jsi.durable, jsi.attached
+func (sub *Subscription) jsiUnsubscribe(jsi *jsSub, drainMode bool) error {
+	sub.mu.Lock()
+	durable, delCons := jsi.durable, jsi.delCons
 	stream, consumer := jsi.stream, jsi.consumer
 	js := jsi.js
 	if jsi.hbc != nil {
 		jsi.hbc.Stop()
 		jsi.hbc = nil
 	}
-	jsi.mu.Unlock()
+	sub.mu.Unlock()
 
-	if drainMode && (durable || attached) {
-		// Skip deleting consumer for durables/attached
-		// consumers when using drain mode.
+	// Delete the JS consumer only if the library created the JS consumer,
```

Comment:
We have been going back and forth here and I am still not convinced we know the answer.

For instance if an app creates a durable and unsubscribes at the end (meaning it wants to do cleanup), but the app gets interrupted and a new version of the app is created, it will not cleanup the durable at the end here.